### PR TITLE
client/authority-discovery: Instrument code with Prometheus

### PR DESCRIPTION
**Pull request by mxinden**
Introduce Prometheus metrics into the authority discovery module
enabling one to observe:

- authority_discovery_times_published_total

- authority_discovery_amount_external_addresses_last_published

- authority_discovery_times_requested_total

- authority_discovery_dht_event_received

---

~~This is still missing a corresponding Polkadot pull request given that one now needs to pass a registry to the authority discovery module.~~

Corresponding Polkadot pull request: https://github.com/paritytech/polkadot/pull/893

### CODE REVIEW

File: client/authority-discovery/src/lib.rs [hunk 1]
```diff
@@ -87,6 +88,53 @@ const LIBP2P_KADEMLIA_BOOTSTRAP_TIME: Duration = Duration::from_secs(30);
...
+	dht_event_received: CounterVec<U64>,
+}
+
+impl Metrics {
+	pub(crate) fn register(registry: &prometheus_endpoint::Registry) -> Result<Self> {
+		Ok(Self {
+			publish: register(
+				Counter::new(
+					"authority_discovery_times_published_total",
+					"Number of time authority discovery published external addresses."
```
Comment (expenses) [inline client/authority-discovery/src/lib.rs]:
```suggestion
					"Number of times the authority discovery has published an external address."
```

File: client/authority-discovery/src/lib.rs [hunk 2]
```diff
@@ -87,6 +88,53 @@ const LIBP2P_KADEMLIA_BOOTSTRAP_TIME: Duration = Duration::from_secs(30);
...
+				)?,
+				registry,
+			)?,
+			amount_last_published: register(
+				Gauge::new(
+					"authority_discovery_amount_external_addresses_last_published",
+					"Amount of external addresses authority discovery published last."
```
Comment (expenses) [inline client/authority-discovery/src/lib.rs]:
```suggestion
					"Number of external addresses published when authority discovery last published addresses ."
```

File: client/authority-discovery/src/lib.rs [hunk 3]
```diff
@@ -87,6 +88,53 @@ const LIBP2P_KADEMLIA_BOOTSTRAP_TIME: Duration = Duration::from_secs(30);
...
+				)?,
+				registry,
+			)?,
+			request: register(
+				Counter::new(
+					"authority_discovery_times_requested_total",
+					"Number of time authority discovery requested external addresses."
```
Comment (expenses) [inline client/authority-discovery/src/lib.rs]:
```suggestion
					"Number of times authority discovery has requested external addresses."
```

**Comment by mxinden** [comment]
Thanks @expenses, appreciated! Can you take another look?

File: client/authority-discovery/src/lib.rs [hunk 4]
```diff
@@ -87,6 +88,53 @@ const LIBP2P_KADEMLIA_BOOTSTRAP_TIME: Duration = Duration::from_secs(30);
...
+	dht_event_received: CounterVec<U64>,
+}
+
+impl Metrics {
+	pub(crate) fn register(registry: &prometheus_endpoint::Registry) -> Result<Self> {
+		Ok(Self {
+			publish: register(
+				Counter::new(
+					"authority_discovery_times_published_total",
+					"Number of times the authority discovery has published external addresses."
```
Comment (expenses) [inline client/authority-discovery/src/lib.rs]: Sorry, there shouldn't have been a `the` there.

```suggestion
					"Number of times authority discovery has published external addresses."
```

**Comment by gavofyork** [comment]
@mxinden tests failing?

**Comment by bkchr** [comment]
@mxinden merging master should fix it.

**Pull request merged**